### PR TITLE
secondlife/viewer#1046 auto exposure quality pass

### PR DESCRIPTION
**Pull request by RunitaiLinden**

### CODE REVIEW

File: indra/newview/pipeline.cpp [hunk 1]
```diff
@@ -6759,6 +6759,8 @@ void LLPipeline::generateLuminance(LLRenderTarget* src, LLRenderTarget* dst)
 
 		gLuminanceProgram.bind();
 
+        static LLCachedControl<F32> diffuse_luminance_scale(gSavedSettings, "RenderDiffuseLuminanceScale", 1.0f);
```
Comment (cosmic-linden) [inline indra/newview/pipeline.cpp:6762]: Not a blocker, but the default value here is different than in settings.xml

File: indra/newview/app_settings/shaders/class1/deferred/luminanceF.glsl [hunk 2]
```diff
@@ -45,7 +47,21 @@ void main()
 {
     vec2 tc = vary_fragcoord*0.6+0.2;
     tc.y -= 0.1; // HACK - nudge exposure sample down a little bit to favor ground over sky
-    vec3 c = texture(diffuseRect, tc).rgb + texture(emissiveRect, tc).rgb;
+    vec3 c = texture(diffuseRect, tc).rgb;
+    
+    vec4  norm         = texture(normalMap, tc);
```
Comment (cosmic-linden) [inline indra/newview/app_settings/shaders/class1/deferred/luminanceF.glsl:52]: What is norm used for?

File: indra/newview/pipeline.cpp [hunk 3]
```diff
@@ -6772,6 +6774,16 @@ void LLPipeline::generateLuminance(LLRenderTarget* src, LLRenderTarget* dst)
 			mGlow[1].bindTexture(0, channel);
 		}
 
+        channel = gLuminanceProgram.enableTexture(LLShaderMgr::DEFERRED_NORMAL);
```
Comment (cosmic-linden) [inline indra/newview/pipeline.cpp:6777]: See above inquiry on norm

File: indra/newview/pipeline.cpp [hunk 4]
```diff
@@ -6772,6 +6774,16 @@ void LLPipeline::generateLuminance(LLRenderTarget* src, LLRenderTarget* dst)
...
+        if (channel > -1)
+        {
+            // bind the normal map to get the environment mask
+            mRT->deferredScreen.bindTexture(2, channel, LLTexUnit::TFO_POINT);
```
Comment (brad-linden) [inline indra/newview/pipeline.cpp:6781]: is this normal mapping change intentional? seems unrelated to this issue

Comment (RunitaiLinden) [reply indra/newview/pipeline.cpp:6781]: yes, the gbuffer normal map w component contains the mask needed to tell if pixels are "sky" pixels or not

File: indra/newview/app_settings/shaders/class1/deferred/luminanceF.glsl [hunk 5]
```diff
@@ -45,7 +47,21 @@ void main()
...
+
+    if (!GET_GBUFFER_FLAG(GBUFFER_FLAG_HAS_HDRI) &&
+        !GET_GBUFFER_FLAG(GBUFFER_FLAG_SKIP_ATMOS))
```
Comment (brad-linden) [inline indra/newview/app_settings/shaders/class1/deferred/luminanceF.glsl:55]: so norm is used by these macros?

**Review by brad-linden — APPROVED** [review]

File: indra/newview/app_settings/shaders/class1/deferred/luminanceF.glsl (hunk 5 above)
Comment (brad-linden) [reply indra/newview/app_settings/shaders/class1/deferred/luminanceF.glsl:55]: confirmed, that is how it is used
